### PR TITLE
Rpp straight-line-pass implementation

### PR DESCRIPTION
**Pull request by ayakayorihiro**
This PR contains new classes and modifications necessary for running the straight-line-pass for RPP.

### CODE REVIEW

File: emop-core/src/main/java/edu/cornell/emop/util/Util.java
Comment (marques-valeria) [inline emop-core/src/main/java/edu/cornell/emop/util/Util.java:140]: Could we call this something besides affectedSpec? So we don't assume RPS or anything

File: emop-maven-plugin/src/main/java/edu/cornell/RppHandlerMojo.java
Comment (marques-valeria) [inline emop-maven-plugin/src/main/java/edu/cornell/RppHandlerMojo.java:56]: So, just for clarification, we're allowing developers to have the ability to choose backgroundSpecs and we'll leave it up to them in case some specs are put in both places or some specs are left out completely? Something that might be nice for the future is that if a spec is in both files it will only be run in critical to avoid redundancy

File: emop-maven-plugin/src/main/java/edu/cornell/RppHandlerMojo.java
Comment (marques-valeria) [inline emop-maven-plugin/src/main/java/edu/cornell/RppHandlerMojo.java:20]: Won't this cause MonitorMojo to execute before RppHandlerMojo? At the moment MonitorMojo is configured entirely for RPS at this seems like it would add a lot of extra overhead at runtime that we don't need

File: emop-maven-plugin/src/main/java/edu/cornell/RppMojo.java
Comment (marques-valeria) [inline emop-maven-plugin/src/main/java/edu/cornell/RppMojo.java:55]: This can be something we add in the vms pr but we should move this function to Utils since it gets used by both vms and rpp

File: emop-maven-plugin/src/main/java/edu/cornell/RppHandlerMojo.java
Comment (ayakayorihiro) [reply emop-maven-plugin/src/main/java/edu/cornell/RppHandlerMojo.java:20]: So I created a change in the `lifecycle.xml` file that stipulates that for the `rpp` lifecycle, the `RppHandlerMojo` will be directly invoked. So I think we won't have any problems where `MonitorMojo` will be invoked before `RppHandlerMojo`.

**Comment by marques-valeria** [comment]
LGTM

File: emop-core/src/main/java/edu/cornell/emop/maven/SurefireMojoInterceptor.java
Comment (ayakayorihiro) [reply emop-core/src/main/java/edu/cornell/emop/maven/SurefireMojoInterceptor.java:46]: Yes! I'm changing the `argLine` field of the surefire mojo :)

File: emop-core/src/main/java/edu/cornell/emop/util/Util.java
Comment (ayakayorihiro) [reply emop-core/src/main/java/edu/cornell/emop/util/Util.java:160]: Sounds good, yeah once we merge both PRs I will probably refactor so we only use one method

**Review by bcarlet — COMMENTED** [review]
Looks good to me.

File: emop-maven-plugin/src/main/java/edu/cornell/MonitorMojo.java
Comment (owolabileg) [inline emop-maven-plugin/src/main/java/edu/cornell/MonitorMojo.java:60]: can you walk me through why the args were removed (I think I understand, but I want to be sure)

File: emop-maven-plugin/src/main/java/edu/cornell/RppHandlerMojo.java
Comment (owolabileg) [inline emop-maven-plugin/src/main/java/edu/cornell/RppHandlerMojo.java:36]: on each --> per

File: emop-maven-plugin/src/main/java/edu/cornell/RppHandlerMojo.java
Comment (owolabileg) [inline emop-maven-plugin/src/main/java/edu/cornell/RppHandlerMojo.java:129]: Do you enforce a line length in your checkstyle config? I usually enforce 120 characters per line

File: emop-maven-plugin/src/main/java/edu/cornell/RppHandlerMojo.java
Comment (owolabileg) [inline emop-maven-plugin/src/main/java/edu/cornell/RppHandlerMojo.java:167]: Add new line at the end. In fact, in a separate PR, you should add a rule to checkstyle to fail if there is no new line at the end

File: emop-maven-plugin/src/main/java/edu/cornell/RppMojo.java
Comment (owolabileg) [inline emop-maven-plugin/src/main/java/edu/cornell/RppMojo.java:59]: skipping --> skip

**Review by owolabileg — CHANGES_REQUESTED** [review]
This is a good first attempt. Well done.

Address the comments and let me know when to take a look again.

File: emop-maven-plugin/src/main/java/edu/cornell/RppHandlerMojo.java
Comment (ayakayorihiro) [reply emop-maven-plugin/src/main/java/edu/cornell/RppHandlerMojo.java:20]: Done!

File: emop-core/src/main/java/edu/cornell/emop/maven/SurefireMojoInterceptor.java
Comment (ayakayorihiro) [reply emop-core/src/main/java/edu/cornell/emop/maven/SurefireMojoInterceptor.java:16]: this was just a note I jotted down when I first learned about this, I'll take it down.

File: emop-core/src/main/java/edu/cornell/emop/maven/SurefireMojoInterceptor.java
Comment (ayakayorihiro) [reply emop-core/src/main/java/edu/cornell/emop/maven/SurefireMojoInterceptor.java:25]: Removed.

File: emop-maven-plugin/src/main/java/edu/cornell/MonitorMojo.java
Comment (ayakayorihiro) [reply emop-maven-plugin/src/main/java/edu/cornell/MonitorMojo.java:60]: The args aren't necessary because the methods called from `generateNewBaseAspect()` that handle the "exclude non affected classes" and "exclude libraries" part both check for the variables `includeNonAffected` and `includeLibraries` on their own.

File: emop-maven-plugin/src/main/java/edu/cornell/RppHandlerMojo.java
Comment (ayakayorihiro) [reply emop-maven-plugin/src/main/java/edu/cornell/RppHandlerMojo.java:129]: There's an enforced line length but it's 125.

File: emop-maven-plugin/src/main/java/edu/cornell/RppHandlerMojo.java
Comment (ayakayorihiro) [reply emop-maven-plugin/src/main/java/edu/cornell/RppHandlerMojo.java:146]: The logic of `computeSpecSets()` makes it such that one can never end up with both `criticalSpecsSet` and `backgroundSpecsSet` being empty (It [exits](https://github.com/SoftEngResearch/emop/blob/312625f4a8c2bbfe63ee07d91e7483c9931bf129/emop-maven-plugin/src/main/java/edu/cornell/RppHandlerMojo.java#L94) if the user/we provide critical and background spec files that are both empty; i.e. for the program to reach this point either of the two sets must be non-empty). `setUpSingleJar` returns an empty string only if the set of specs to monitor is empty. So, the non-emptiness of `criticalRunAgentPath` imply the non-emptiness of `backgroundRunAgentPath`.

File: emop-maven-plugin/src/main/java/edu/cornell/RppHandlerMojo.java
Comment (ayakayorihiro) [reply emop-maven-plugin/src/main/java/edu/cornell/RppHandlerMojo.java:167]: why is it a good practice to add a newline at the end?

File: emop-maven-plugin/src/main/java/edu/cornell/RppMojo.java
Comment (ayakayorihiro) [reply emop-maven-plugin/src/main/java/edu/cornell/RppMojo.java:20]: Fixed this one.

File: emop-core/src/main/java/edu/cornell/emop/maven/SurefireMojoInterceptor.java
Comment (owolabileg) [inline emop-core/src/main/java/edu/cornell/emop/maven/SurefireMojoInterceptor.java:31]: Is there a more specific exception that should be thrown here?

File: emop-core/src/main/java/edu/cornell/emop/maven/SurefireMojoInterceptor.java
Comment (owolabileg) [inline emop-core/src/main/java/edu/cornell/emop/maven/SurefireMojoInterceptor.java:26]: the comment should start on the same column as the previous statement

File: emop-core/src/main/java/edu/cornell/emop/maven/SurefireMojoInterceptor.java
Comment (owolabileg) [inline emop-core/src/main/java/edu/cornell/emop/maven/SurefireMojoInterceptor.java:44]: add empty line at the end of the file

File: emop-core/src/main/java/edu/cornell/emop/util/Util.java
Comment (owolabileg) [reply emop-core/src/main/java/edu/cornell/emop/util/Util.java:78]: Please address this before merging

**Review by owolabileg — COMMENTED** [review]
Address the outstanding review comments and then you can merge.

File: emop-maven-plugin/src/main/java/edu/cornell/RppHandlerMojo.java
Comment (ayakayorihiro) [reply emop-maven-plugin/src/main/java/edu/cornell/RppHandlerMojo.java]: Interesting; thanks!

**Pull request merged**
